calculator_test.go

```diff
@@ -2,23 +2,195 @@ package calculator_test
+	s := rand.NewSource(time.Now().UnixNano())
+	r := rand.New(s)
+
+	a := r.Float64() * float64(r.Intn(10000))
+	b := r.Float64() * float64(r.Intn(10000))
+	return []float64{a, b}
+}
+
 func TestAdd(t *testing.T) {
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+
+	tt := []struct {
+		name        string
+		nums        []float64
+		want        float64
+		expectedErr bool
+	}{
+		{name: "Add two positive numbers", nums: []float64{1, 2}, want: 3, expectedErr: false},
+		{name: "Add two negative numbers", nums: []float64{-4.5, -5.5}, want: -10, expectedErr: false},
+		{name: "Add multiple numbers", nums: []float64{-4.5, -5.5, 10, 2.5}, want: 2.5, expectedErr: false},
+		{name: "Add no numbers", nums: []float64{}, want: 0, expectedErr: true},
+		{name: "Add one number", nums: []float64{10}, want: 10, expectedErr: true},
+	}
+
+	for _, tc := range tt {
+		got, err := calculator.Add(tc.nums...)
+
+		if err != nil && !tc.expectedErr {
+			t.Errorf("%s Add(%v) should return an error", tc.name, tc.nums)
+		}
+
+		if tc.want != got && !tc.expectedErr {
+			t.Errorf("%s Add(%v) = %f, want %f", tc.name, tc.nums, got, tc.want)
+		}
+	}
+}
+
+func TestAddRandom(t *testing.T) {
+	t.Parallel()
+
+	name := "Add randomly generated numbers"
+
+	for i := 0; i < 100; i++ {
+		a := GenerateNums()
+		want := a[0] + a[1]
+
+		// Do not check err here as in this test
+		// we always pass two params to the function.
+		got, _ := calculator.Add(a...)
+
+		if got != want {
+			t.Errorf("%s, Add(%v) = %f, want: %f", name, a, got, want)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+
+	tt := []struct {
+		name        string
+		nums        []float64
+		want        float64
+		expectedErr bool
+	}{
+		{name: "Substract two positive numbers", nums: []float64{4, 2}, want: 2, expectedErr: false},
+		{name: "Substract two negative numbers", nums: []float64{-4.5, -5.5}, want: 1, expectedErr: false},
+		{name: "Substract from zero", nums: []float64{0, 5.45}, want: -5.45, expectedErr: false},
+		{name: "Substract multiple numbers", nums: []float64{20, 5.5, 4, 3, 1.25}, want: 6.25, expectedErr: false},
+		{name: "Substract incorrect number of params", nums: []float64{4}, want: 0, expectedErr: true},
+		{name: "Substract incorrect number of params", nums: []float64{}, want: 0, expectedErr: true},
+	}
+
+	for _, tc := range tt {
+		got, err := calculator.Subtract(tc.nums...)
+
+		if err != nil && !tc.expectedErr {
+			t.Errorf("%s, Substract(%v) should return an error", tc.name, tc.nums)
+		}
+
+		if tc.want != got && !tc.expectedErr {
+			t.Errorf("%s, Substract(%v) = %f, want %f", tc.name, tc.nums, got, tc.want)
+		}
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestMultiply(t *testing.T) {
```

Comment:
An interesting thought is whether (since the test case structure is the same) we could test Add, Subtract, and Multiply all in one test! We would need some way for the test case struct to control which function was called, wouldn't we?